File: extension/android/executorch_android/src/main/java/org/pytorch/executorch/ExecutorchRuntimeException.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

package org.pytorch.executorch;

import com.facebook.jni.annotations.DoNotStrip;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * Base exception for all ExecuTorch runtime errors. Each instance carries an integer error code
 * corresponding to the native {@code runtime/core/error.h} values, accessible via {@link
 * #getErrorCode()}.
 */
public class ExecutorchRuntimeException extends RuntimeException {
  // Error code constants - keep in sync with runtime/core/error.h

  // System errors

  /** Operation completed successfully. */
  public static final int OK = 0x00;

  /** An unexpected internal error occurred in the runtime. */
  public static final int INTERNAL = 0x01;

  /** The runtime or method is in an invalid state for the requested operation. */
  public static final int INVALID_STATE = 0x02;

  /** The method has finished execution and has no more work to do. */
  public static final int END_OF_METHOD = 0x03;

  /** A required resource has already been loaded. */
  public static final int ALREADY_LOADED = 0x04;

  // Logical errors

  /** The requested operation is not supported by this build or backend. */
  public static final int NOT_SUPPORTED = 0x10;

  /** The requested operation has not been implemented. */
  public static final int NOT_IMPLEMENTED = 0x11;

  /** One or more arguments passed to the operation are invalid. */
  public static final int INVALID_ARGUMENT = 0x12;

  /** A value or tensor has an unexpected type. */
  public static final int INVALID_TYPE = 0x13;

  /** A required operator kernel is not registered. */
  public static final int OPERATOR_MISSING = 0x14;

  /** The maximum number of registered kernels has been exceeded. */
  public static final int REGISTRATION_EXCEEDING_MAX_KERNELS = 0x15;

  /** A kernel with the same name is already registered. */
  public static final int REGISTRATION_ALREADY_REGISTERED = 0x16;

  // Resource errors

  /** A required resource (file, tensor, program) was not found. */
  public static final int NOT_FOUND = 0x20;

  /** A memory allocation failed. */
  public static final int MEMORY_ALLOCATION_FAILED = 0x21;

  /** Access to a resource was denied or failed. */
  public static final int ACCESS_FAILED = 0x22;

  /** The loaded program is malformed or incompatible. */
  public static final int INVALID_PROGRAM = 0x23;

  /** External data referenced by the program is invalid or missing. */
  public static final int INVALID_EXTERNAL_DATA = 0x24;

  /** The system has run out of a required resource. */
  public static final int OUT_OF_RESOURCES = 0x25;

  // Delegate errors

  /** A delegate reported an incompatible model or configuration. */
  public static final int DELEGATE_INVALID_COMPATIBILITY = 0x30;

  /** A delegate failed to allocate required memory. */
  public static final int DELEGATE_MEMORY_ALLOCATION_FAILED = 0x31;

  /** A delegate received an invalid or stale handle. */
  public static final int DELEGATE_INVALID_HANDLE = 0x32;

  private static final Map<Integer, String> ERROR_CODE_MESSAGES;

  static {
    Map<Integer, String> map = new HashMap<>();

    // System errors
    map.put(OK, "Operation successful");
    map.put(INTERNAL, "Internal error");
    map.put(INVALID_STATE, "Invalid state");
    map.put(END_OF_METHOD, "End of method reached");
    map.put(ALREADY_LOADED, "Already loaded");
    // Logical errors
    map.put(NOT_SUPPORTED, "Operation not supported");
    map.put(NOT_IMPLEMENTED, "Operation not implemented");
    map.put(INVALID_ARGUMENT, "Invalid argument");
    map.put(INVALID_TYPE, "Invalid type");
    map.put(OPERATOR_MISSING, "Operator missing");
    map.put(REGISTRATION_EXCEEDING_MAX_KERNELS, "Exceeded max kernels");
    map.put(REGISTRATION_ALREADY_REGISTERED, "Kernel already registered");
    // Resource errors
    map.put(NOT_FOUND, "Resource not found");
    map.put(MEMORY_ALLOCATION_FAILED, "Memory allocation failed");
    map.put(ACCESS_FAILED, "Access failed");
    map.put(INVALID_PROGRAM, "Invalid program");
    map.put(INVALID_EXTERNAL_DATA, "Invalid external data");
    map.put(OUT_OF_RESOURCES, "Out of resources");
    // Delegate errors
    map.put(DELEGATE_INVALID_COMPATIBILITY, "Delegate invalid compatibility");
    map.put(DELEGATE_MEMORY_ALLOCATION_FAILED, "Delegate memory allocation failed");
    map.put(DELEGATE_INVALID_HANDLE, "Delegate invalid handle");
    ERROR_CODE_MESSAGES = Collections.unmodifiableMap(map);
  }

  static class ErrorHelper {
    static String formatMessage(int errorCode, String details) {
      String baseMessage = ERROR_CODE_MESSAGES.get(errorCode);
      if (baseMessage == null) {
        baseMessage = "Unknown error code 0x" + Integer.toHexString(errorCode);
      }

      String safeDetails = details != null ? details : "No details provided";
      return String.format(
          "[ExecuTorch Error 0x%s] %s: %s",
          Integer.toHexString(errorCode), baseMessage, safeDetails);
    }

    static String getDetailedErrorLogs() {
      StringBuilder sb = new StringBuilder();
      try {
        String[] logEntries = Module.readLogBufferStatic(); // JNI call
        if (logEntries != null && logEntries.length > 0) {
          sb.append("\nDetailed logs:\n");
          for (String entry : logEntries) {
            sb.append(entry).append("\n");
          }
        }
      } catch (Exception e) {
        sb.append("Failed to retrieve detailed logs: ").append(e.getMessage());
      }
      return sb.toString();
    }
  }

  private final int errorCode;

  @DoNotStrip
  public ExecutorchRuntimeException(int errorCode, String details) {
    super(ErrorHelper.formatMessage(errorCode, details));
    this.errorCode = errorCode;
  }

  public ExecutorchRuntimeException(int errorCode, String details, Throwable cause) {
    super(ErrorHelper.formatMessage(errorCode, details), cause);
    this.errorCode = errorCode;
  }

  /** Returns the numeric error code from {@code runtime/core/error.h}. */
  public int getErrorCode() {
    return errorCode;
  }

  /** Returns detailed log output captured from the native runtime, if available. */
  public String getDetailedError() {
    return ErrorHelper.getDetailedErrorLogs();
  }

  @DoNotStrip
  public static class ExecutorchInvalidArgumentException extends ExecutorchRuntimeException {
    @DoNotStrip
    public ExecutorchInvalidArgumentException(String details) {
      super(INVALID_ARGUMENT, details);
    }
  }

  @DoNotStrip
  public static RuntimeException makeExecutorchException(int errorCode, String details) {
    switch (errorCode) {
      case INVALID_ARGUMENT:
        return new ExecutorchInvalidArgumentException(details);
      default:
        return new ExecutorchRuntimeException(errorCode, details);
    }
  }
}

File: extension/java/src/main/java/org/pytorch/executorch/ExecutorchRuntimeException.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

package org.pytorch.executorch

import com.facebook.jni.annotations.DoNotStrip

/**
 * Base exception for all ExecuTorch runtime errors. Each instance carries an integer error code
 * corresponding to the native `runtime/core/error.h` values, accessible via [getErrorCode].
 */
open class ExecutorchRuntimeException
@DoNotStrip
constructor(
    val errorCode: Int,
    details: String?,
) : RuntimeException(ErrorHelper.formatMessage(errorCode, details)) {

  constructor(
      errorCode: Int,
      details: String?,
      cause: Throwable?,
  ) : this(errorCode, details) {
    if (cause != null) initCause(cause)
  }

  /** Returns detailed log output captured from the native runtime, if available. */
  fun getDetailedError(): String = ErrorHelper.getDetailedErrorLogs()

  @DoNotStrip
  class ExecutorchInvalidArgumentException @DoNotStrip constructor(details: String?) :
      ExecutorchRuntimeException(INVALID_ARGUMENT, details)

  private object ErrorHelper {
    private val ERROR_CODE_MESSAGES: Map<Int, String> =
        mapOf(
            // System errors
            OK to "Operation successful",
            INTERNAL to "Internal error",
            INVALID_STATE to "Invalid state",
            END_OF_METHOD to "End of method reached",
            ALREADY_LOADED to "Already loaded",
            // Logical errors
            NOT_SUPPORTED to "Operation not supported",
            NOT_IMPLEMENTED to "Operation not implemented",
            INVALID_ARGUMENT to "Invalid argument",
            INVALID_TYPE to "Invalid type",
            OPERATOR_MISSING to "Operator missing",
            REGISTRATION_EXCEEDING_MAX_KERNELS to "Exceeded max kernels",
            REGISTRATION_ALREADY_REGISTERED to "Kernel already registered",
            // Resource errors
            NOT_FOUND to "Resource not found",
            MEMORY_ALLOCATION_FAILED to "Memory allocation failed",
            ACCESS_FAILED to "Access failed",
            INVALID_PROGRAM to "Invalid program",
            INVALID_EXTERNAL_DATA to "Invalid external data",
            OUT_OF_RESOURCES to "Out of resources",
            // Delegate errors
            DELEGATE_INVALID_COMPATIBILITY to "Delegate invalid compatibility",
            DELEGATE_MEMORY_ALLOCATION_FAILED to "Delegate memory allocation failed",
            DELEGATE_INVALID_HANDLE to "Delegate invalid handle",
        )

    fun formatMessage(errorCode: Int, details: String?): String {
      val baseMessage =
          ERROR_CODE_MESSAGES[errorCode] ?: "Unknown error code 0x${Integer.toHexString(errorCode)}"
      val safeDetails = details ?: "No details provided"
      return "[ExecuTorch Error 0x${Integer.toHexString(errorCode)}] $baseMessage: $safeDetails"
    }

    fun getDetailedErrorLogs(): String {
      val sb = StringBuilder()
      try {
        val logEntries = Module.readLogBufferStatic() // JNI call
        if (logEntries != null && logEntries.isNotEmpty()) {
          sb.append("\nDetailed logs:\n")
          for (entry in logEntries) {
            sb.append(entry).append("\n")
          }
        }
      } catch (e: Exception) {
        return ""
      }
      return sb.toString()
    }
  }

  companion object {
    // Error code constants - keep in sync with runtime/core/error.h

    // System errors
    const val OK = 0x00
    const val INTERNAL = 0x01
    const val INVALID_STATE = 0x02
    const val END_OF_METHOD = 0x03
    const val ALREADY_LOADED = 0x04

    // Logical errors
    const val NOT_SUPPORTED = 0x10
    const val NOT_IMPLEMENTED = 0x11
    const val INVALID_ARGUMENT = 0x12
    const val INVALID_TYPE = 0x13
    const val OPERATOR_MISSING = 0x14
    const val REGISTRATION_EXCEEDING_MAX_KERNELS = 0x15
    const val REGISTRATION_ALREADY_REGISTERED = 0x16

    // Resource errors
    const val NOT_FOUND = 0x20
    const val MEMORY_ALLOCATION_FAILED = 0x21
    const val ACCESS_FAILED = 0x22
    const val INVALID_PROGRAM = 0x23
    const val INVALID_EXTERNAL_DATA = 0x24
    const val OUT_OF_RESOURCES = 0x25

    // Delegate errors
    const val DELEGATE_INVALID_COMPATIBILITY = 0x30
    const val DELEGATE_MEMORY_ALLOCATION_FAILED = 0x31
    const val DELEGATE_INVALID_HANDLE = 0x32

    @DoNotStrip
    @JvmStatic
    fun makeExecutorchException(errorCode: Int, details: String?): RuntimeException {
      val nativeTail =
          try {
            ErrorHelper.getDetailedErrorLogs()
                .removePrefix("\nDetailed logs:\n")
                .replace(Regex("\\s+"), " ")
                .trim()
          } catch (t: Throwable) {
            ""
          }
      val enrichedDetails =
          if (nativeTail.isNotBlank()) {
            "${details ?: "No details provided"} | nativeLog=${nativeTail.takeLast(NATIVE_LOG_TAIL_MAX_CHARS)}"
          } else {
            details
          }
      return when (errorCode) {
        INVALID_ARGUMENT -> ExecutorchInvalidArgumentException(enrichedDetails)
        else -> ExecutorchRuntimeException(errorCode, enrichedDetails)
      }
    }

    private const val NATIVE_LOG_TAIL_MAX_CHARS = 2048
  }
}
